### ISSUE TITLE
Building an Android app where heavy lifting is done by backend

### ISSUE THREAD

**Issue by jonathanknowles** [obsidiansystems/obelisk#348]
If I have understood correctly, Obelisk makes it possible to build an **Reflex/Android** app with a frontend (compiled with GHCJS) and a backend (compiled with GHC).

This sounds as though it might be perfect for a use case I have in mind:

## Use Case

Imagine we want to provide a UI for the following two functions:

* `wibble :: Int -> Int` (takes around 10 milliseconds to execute with typical input if compiled with GHC)
* `wobble :: Int -> Int` (takes around 5 seconds to execute with typical input if compiled with GHC).

For performance reasons, it's important that the functions are compiled with **GHC**, so they would be located in the **backend**. (Perhaps they perform some heavy data processing, or perhaps they must interact with a file-based database.)

Imagine then that we wish to provide the following UI **frontend** (defined with Reflex DOM):
```
input:  [..............] <-- accepts an integer input
button: [compute wibble] <-- initiates a call to function "wibble" in the backend
button: [compute wobble] <-- initiates a call to function "wobble" in the backend
result: [..............] <-- result eventually appears here
```

In other words: the user asks the frontend to initiate a computation. The frontend then calls the backend with a description of the computation required (either `wibble` or `wobble`). Once the backend has computed the result, it passes the result back to the frontend, which displays the result for the user.

## Question

Is Obelisk suitable for this use case?

If so, how should we go about wiring this up? Would `servant` be an appropriate tool to use here (for defining the API between frontend and backend), or is it possible to build with the built-in routing functionality provided by Obelisk?

Perhaps someone has already built an example of something similar?

If not, then I would be interested in building this as a demo project. (Perhaps it might make an interesting blog article to help others who are interested.)

**Comment by erikryb** [comment]
Did you manage to do this? I have a similar situation. It seems simple to do, but it would help if there were any example of this.

**Comment by alexfmpe** [comment]
> If I have understood correctly, Obelisk makes it possible to build an Reflex/Android app with a frontend (compiled with GHCJS) and a backend (compiled with GHC).

I believe the Android code is also compiled with GHC, and the javascript involved, if any, is only for rendering. If you run `nix-build -A android.frontend`, the build logs are always for 'ghc' not 'ghcjs'
AFAICT GHCJS is only used for webapps.

**Comment by 3noch** [comment]
I missed this so I apologise for not answering sooner! Yes Obelisk is a perfect tool for this. It has a routing module that lets you define requests made from the frontend to the backend. We are working on better documentation for that as well. Feel free to keep asking questions here.

**Comment by erikryb** [comment]
Great! Any simple examples of doing requests from frontend to backend, such as the use-case of OP? Would it be best to use servant for this, or is it easier to use the routing module of Obelisk?

**Comment by srid** [comment]
The general workflow I've have used is as follows:

1. Define your routes in `common/src/Route.hs`
2. Handle them in the backend in `backend/src/Backend.hs` (have it return JSON)
3. Use Reflex.Dom.Xhr (eg: `getAndDecode`) in the frontend along with `checkEncoder` to make requests

Here's an app that does all that https://github.com/srid/Taut (although I suggest ignoring the `PaginatedRoute` part which is more of a hack).

**Comment by erikryb** [comment]
I am now starting to understand the routing system in Obelisk, but I have one problem. My app works when I do `ob run`, but when I try to deploy it to Android with `nix-build -A android.frontend -o result-android`, the frontend no longer recieves anything from the backend. Is the backend compiled at all, or do I have to use another command for deploying both frontend and backend to Android?

**Comment by srid** [comment]
Mobile apps are built only from the frontend (not backend). Typically you would deploy your backend to a remote server (`ob deploy`), and then configure the `config/common/route` of your mobile build to point to it.

I don't know how exactly the `ob deploy test android` command relates (if at all) to any of this, though.

**Comment by erikryb** [comment]
Ahh, I had the wrong picture in my mind then. I thought the frontend was compiled to js, and the backend to native, and that they both ran on the mobile, but then I see from Alex' comment above that the frontend is also compiled to native. Thanks!

**Comment by erikryb** [comment]
> Mobile apps are built only from the frontend (not backend). Typically you would deploy your backend to a remote server (`ob deploy`), and then configure the ~`config/common/route`~ API route config of your mobile build to point to it.

I am trying to do some testing using my nixos laptop as server, but I am having some trouble. The `config/common/route` file contains `https://192.168.0.2:8000`, which is the local IP of my nixos server. I have then made an android build using `nix-build -A android.frontend -o result-android`. I am now running `ob run` on the server, but the android app is not able to connect to it. I have also put

```
networking.firewall.allowedTCPPorts = [8000];
networking.firewall.allowedUDPPorts = [8000];
```

in my `configuration.nix`. I would greatly appreciate some help.

**Comment by ali-abrar** [comment]
Try http instead of https

**Comment by erikryb** [comment]
Still not working with `http`. I am able to open the web version of the app when I type `http://192.168.0.2` in the browser of the Android phone, but the compiled Android app is unable to connect to the server.

**Comment by 3noch** [comment]
How are you compiling the android app? The android app needs to have the remote URL configured and that's typically done with `ob deploy` tooling.

**Comment by erikryb** [comment]
Ahh, I just compiled it using `nix-build -A android.frontend -o result-android`. I will give `ob deploy` a try.

**Comment by erikryb** [comment]
I have tried using `ob deploy` and `ob deploy test android`, but still not working. In the apk generated by `ob deploy test android` there is the file `assets/config/common/route` which contains `http://192.168.0.2:8000`. I have also tried with `https://192.168.0.2:8000`. How should I debug this? Is there a way to monitor the http requests made by the android app?

**Comment by 3noch** [comment]
@erikryb I've not tested trying to connect an APK app to `ob run`. Normally I've built `nix-build -A exe` (a full build, see the README) and run the server locally on my laptop that way. But before that, are you able to access the app from your Android phone's web browser?

**Comment by erikryb** [comment]
Yes, I am able to access it from the phone's web browser, so there should be nothing wrong about the connection. So you are suggesting using `nix-build -A exe` to build the backend? Should I still use `ob deploy init` and `ob deploy test android` to build the Android app? I tried running `nix-build -A exe` but the ghcjs compiler failed to build some of the dependencies of the frontend. I could try to fix that, but I'm not sure if I really need the ghcjs compiled frontend. All I need is the android compiled frontend and the ghc compiled backend (I think).

**Comment by alexfmpe** [comment]
> Yes, I am able to access it from the phone's web browser, so there should be nothing wrong about the connection.  

Could the issue be related to missing app permissions?

**Comment by 3noch** [comment]
@erikryb I see. Obelisk doesn't *officially* support this precise use case yet. Without seeing your code, my best recommendation for now is neuter your frontend on GHCJS with CPP so that `nix-build -A exe` works.

In your cabal file you can make those dependencies rely on the compiler not being GHCJS: https://github.com/reflex-frp/reflex-dom/blob/9c696a0f382321c72237c206699efda2e30faef4/reflex-dom/reflex-dom.cabal#L75

and/or CPP certain features away: https://github.com/reflex-frp/reflex-dom/blob/9c696a0f382321c72237c206699efda2e30faef4/reflex-dom/src/Reflex/Dom/Internal.hs#L26

**Comment by erikryb** [comment]
I have done some debugging now, and I think I managed to isolate the problem. For context, I have the following function in `Frontend.hs`:

```
fetchSomeText ::
   (MonadHold t m, PostBuild t m, Prerender js t m) => m (Event t (Maybe Text))
 fetchSomeText = fmap switchDyn $ prerender (pure never) $ do
   pb <- getPostBuild
   getAndDecode $ renderBackendRoute enc (BackendRoute_ServeSomeText :/ ()) <$ pb
   where
     Right (enc :: Encoder Identity Identity (R (FullRoute BackendRoute FrontendRoute)) PageName) = checkEncoder fullRouteEncoder
```

This function is basically the same as `getSearchExamples` from [here](https://github.com/srid/Taut/blob/master/frontend/src/Frontend/Message.hs) but with a different return type. I have also modified the routes in `Common/Route.hs`:

```
 data BackendRoute :: * -> * where
   -- | Used to handle unparseable routes.
   BackendRoute_Missing :: BackendRoute ()
   BackendRoute_ServeSomeText :: BackendRoute ()
   -- You can define any routes that will be handled specially by the backend here.
   -- i.e. These do not serve the frontend, but do something different, such as serving static files.

 fullRouteEncoder
   :: Encoder (Either Text) Identity (R (FullRoute BackendRoute FrontendRoute)) PageName
 fullRouteEncoder = mkFullRouteEncoder
   (FullRoute_Backend BackendRoute_Missing :/ ())
   (\case
       BackendRoute_Missing -> PathSegment "missing" $ unitEncoder mempty
       BackendRoute_ServeSomeText -> PathSegment "serveSomeText" $ unitEncoder mempty)
   (\case
       FrontendRoute_Main -> PathEnd $ unitEncoder mempty)
```

In the `Backend.hs` I have the following:

```
 backend :: Backend BackendRoute FrontendRoute
 backend = Backend
   { _backend_run = \serve -> do
       serve $ \case
         BackendRoute_Missing :=> Identity () -> return ()
         BackendRoute_ServeSomeText :=> Identity () -> do
           writeLBS $ encode ("Some text" :: Text)
   , _backend_routeEncoder = fullRouteEncoder
   }
```

Now, it seems that the code `renderBackendRoute enc (BackendRoute_ServeSomeText :/ ()) <$ pb` in `fetchSomeText` returns the url `"/serveSomeText"` without the IP address. This happens both when I run locally with `ob run`, and also on the Android build (which I figured out after some debugging using Android Studio). This doesn't seem like the correct behaviour. Shouldn't this function somehow get the server IP from `config/common/route`? One workaround that works for me is to manually add the IP like this:

```
getAndDecode $ ("http://192.168.0.2:8000" <> renderBackendRoute enc (BackendRoute_ServeSomeText :/ ()) <$ pb)
```

This workaround works on the Android, and I am able to fetch the text from my server running just `ob run`, but I doubt this is the official way to do this.

To summarize: As it turned out, it seems like the problem lies with the way the Android build is compiled, both with `ob deploy test android` and the `nix-build -A android.frontend -o result-android`. It also turned out that once the problem is fixed, either with my workaround, or the "official way" (which I haven't figured out yet), the Android build is able to communicate with the server running `ob run`, without needing to do `ob develop` or `nix-build -A exe` (but thanks anyway @3noch, it's nice to know).

**Comment by 3noch** [comment]
@erikryb Wow thank you so much for this detailed writeup! And what you're saying makes complete sense. This is *definitely* an issue we need to solve so that no one runs into this again! So we'll use this ticket as a placeholder for that until we have it solved.

**Comment by 3noch** [comment]
I'm going to close this as the original question has been answered.
